r/R/dataset.R

```diff
@@ -64,6 +64,9 @@ open_dataset <- function(sources,
                          partitioning = hive_partition(),
                          unify_schemas = NULL,
                          ...) {
+  if (isFALSE(inherits(schema, "Schema")) && isFALSE(is.null(schema))) {
+    stop("'schema' must be a Schema object", call. = FALSE)
+  }
```

Comment:
I think the better place for this check is lower down, in DatasetFactory$Finish. Add `assert_is(schema, "Schema")` before L30 in dataset-factory.R

The reasoning is that since we're trying to prevent a segfault in the C++ code, we should put the validation as close to the place where we get into the C++ code. That way, if someone were to be using the lower-level interface to the C++ library, they're still protected from these cases.

Reply:
By the same reasoning, we should `assert_is(schema, "Schema")` on L80 as well because there is a different C++ entry point on L85.

Reply:
all set ! locally, it passes the tests